vignettes/join-keys.Rmd

```diff
@@ -49,51 +109,102 @@ data_1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)
-columns are present in the parent table.
+When passing multiple datasets to the `teal_data` function, dataset relationship are set using `join_keys` and `join_key` and these are used to merge datasets together within `teal` apps.
+For users familiar with `SQL` database schema, these relationships are symmetric and not as strict as `SQL` foreign key relationships as `teal` does not validate whether the values inserted into foreign key columns are present in the parent table.
 
 For example:
 
-```{r, eval=FALSE}
+```{r}
 library(teal.data)
 
 data_1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)
 data_2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))
 data_3 <- data.frame(V = factor(5:14), T = 4)
 
 data <- teal_data(
-  dataset("D1", data_1, code = "D1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)"),
-  dataset("D2", data_2, code = "D2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))"),
-  dataset("D3", data_3, code = "D3 <- data.frame(V = factor(5:14), T = 4)"),
+  D1 = data_1,
+  D2 = data_2,
+  D3 = data_3,
   join_keys = join_keys(
     join_key("D1", "D2", c("X" = "W")),
     join_key("D2", "D3", c("V" = "V"))
+  ),
+  code = c(
+    "D1 <- data.frame(X = factor(1:10), Y = 21:30, Z = 1:10)",
+    "D2 <- data.frame(W = factor(10:1), V = factor(5:14), M = rep(1:5, 2))",
+    "D3 <- data.frame(V = factor(5:14), T = 4)"
   )
 )
+
+join_keys(data)
+```
+
+### Implicit relationships
+
+This is a special inferred relationship from existing `join_keys` that does not need to be explicitly defined.
+If two datasets that are a child of the same parent _(and have common keys)_, then there is an implicit relationship between them.
+
+These implicit relationships can be used to merge 2 datasets together, just as if they were defined manually.
+
+```{r}
+jk <- join_keys(
+  join_key("ds1", keys = "id1"),
+  join_key("ds2", keys = c("id1", "id2")),
+  join_key("ds3", keys = c("id1", "id3")),
+  join_key("ds1", "ds2", keys = "id1"),
+  join_key("ds1", "ds3", keys = "id1")
+)
+
+# The parent of ds2 and ds3 is ds1
+parents(jk) <- list(ds2 = "ds1", ds3 = "ds1")
+
+jk["ds3", "ds2"]
 ```
 
-The `join_keys` function is used to specify keys:
+_note_: the definition above contains no `join_key` for `"ds2" <-> "ds3"`
+
+### `ADaM` `join_keys`
+
+When working with datasets named according to the `ADaM` standard, `teal.data` can assign the keys on its own using the `teal.data::default_cdisc_join_keys` that contains primary and foreign keys for standard datasets (`r paste("\u0060", names(default_cdisc_join_keys), "\u0060", collapse = ", ", sep = "")`).
+
+The full set of join keys can be used, or a subset of those.
+Please note, that if the subsetted dataset has a parent, that will also be returned.
+
+```{r}
+default_cdisc_join_keys["ADSL"]
+
+default_cdisc_join_keys["ADTTE"]
+
+default_cdisc_join_keys[c("ADSL", "ADTTE", "ADRS")]
+```
+
+#### The full set of `ADaM` `join_keys`
+
+```{r}
+default_cdisc_join_keys
+```
```

Comment:
I've changed the order, but didn't add `echo=FALSE` as that will remove any context of what the user is seeing.

However, I don't think that moving the full `default_cdisc_join_keys` to top will have any benefit as it will decrease the value of the other examples.

I won't die on this though :sweat_smile:

Reply:
Im fin with current state already. I had the biggest issue with data in the examples, it looks good now